src/action.c

```diff
@@ -356,23 +364,75 @@ actions_contain_toggle_keybinds(struct wl_list *action_list)
 	return false;
 }
 
-void action_list_free(struct wl_list *action_list)
+/* Checks for *required* arguments */
+bool
+action_is_valid(struct action *action)
 {
+	char *arg_name = NULL;
+	switch (action->type) {
+	case ACTION_TYPE_EXECUTE:
+		arg_name = xstrdup("command");
+		break;
+	case ACTION_TYPE_MOVE_TO_EDGE:
+	case ACTION_TYPE_SNAP_TO_EDGE:
+		arg_name = xstrdup("direction");
+		break;
+	case ACTION_TYPE_SHOW_MENU:
+		arg_name = xstrdup("menu");
+		break;
+	case ACTION_TYPE_GO_TO_DESKTOP:
+	case ACTION_TYPE_SEND_TO_DESKTOP:
+		arg_name = xstrdup("to");
+		break;
+	case ACTION_TYPE_SNAP_TO_REGION:
+		arg_name = xstrdup("region");
+		break;
+	case ACTION_TYPE_FOCUS_OUTPUT:
+		arg_name = xstrdup("output");
+		break;
+	default:
+		/* No arguments required */
+		return true;
+	}
+
+	if (arg_value_exists(action, arg_name)) {
+		free(arg_name);
+		return true;
+	}
+
+	wlr_log(WLR_ERROR, "Missing required argument for %s: %s",
+		action_names[action->type], arg_name);
+	free(arg_name);
+	return false;
+}
+
+void
+action_free(struct action *action)
+{
+	/* Free args */
 	struct action_arg *arg, *arg_tmp;
+	wl_list_for_each_safe(arg, arg_tmp, &action->args, link) {
+		wl_list_remove(&arg->link);
+		zfree(arg->key);
+		if (arg->type == LAB_ACTION_ARG_STR) {
+			/*
+			 * struct action_arg_str *str_arg = (struct action_arg_str *)arg;
+			 * zfree(str_arg->value);
+			 */
+			zfree(((struct action_arg_str *)arg)->value);
+		}
```

Comment:
@johanmalm which of the versions do you prefer, the commented or the other one?
I actually slightly prefer the commented 2 line variant as its easier to read.

Reply:
Yes, agree.
I prefer the commented out version too.

Reply:
done